db/analysis.go

```diff
@@ -0,0 +1,124 @@
+package db
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/jackc/pgx/v4/pgxpool"
+	"github.com/likecoin/likecoin-chain-tx-indexer/logger"
+)
+
+func GetNftCount(conn *pgxpool.Conn, q QueryNftCountRequest) (count QueryCountResponse, err error) {
+	sql := `
+	SELECT COUNT(n.id)
+	FROM nft as n
+	JOIN nft_class AS c USING (class_id)
+	JOIN iscn as i ON c.parent_iscn_id_prefix = i.iscn_id_prefix
+	WHERE ($1 = true OR i.owner != n.owner)
+		AND ($2::text[] IS NULL OR n.owner != ALL($2))
+	`
+	ctx, cancel := GetTimeoutContext()
+	defer cancel()
+
+	err = conn.QueryRow(ctx, sql, q.IncludeOwner, q.IgnoreList).Scan(&count.Count)
+	if err != nil {
+		err = fmt.Errorf("get nft count failed: %w", err)
+		logger.L.Error(err, q)
+	}
+	return
+}
+
+func GetNftTradeStats(conn *pgxpool.Conn, q QueryNftTradeStatsRequest) (res QueryNftTradeStatsResponse, err error) {
+	payload := struct {
+		Type    string `json:"@type"`
+		Grantee string `json:"grantee"`
+	}{
+		"/cosmos.authz.v1beta1.MsgExec",
+		q.ApiAddress,
+	}
+	payloadJSON, err := json.Marshal(&payload)
+	if err != nil {
+		err = fmt.Errorf("marshal nft trade stats payload failed: %w", err)
+		logger.L.Error(err, q)
+		return
+	}
+
+	sql := `
+	SELECT count(*), sum((tx #>> '{"tx", "body", "messages", 0, "msgs", 0, "amount", 0, "amount"}')::bigint)
+	from txs
+	WHERE tx #> '{"tx", "body", "messages", 0}' @> $1
+	`
+	ctx, cancel := GetTimeoutContext()
+	defer cancel()
+	err = conn.QueryRow(ctx, sql, string(payloadJSON)).Scan(
+		&res.Count, &res.TotalVolume,
+	)
+	if err != nil {
+		err = fmt.Errorf("get nft trade stats failed: %w", err)
+		logger.L.Error(err, q)
+	}
+	return
+}
+
+func GetNftCreatorCount(conn *pgxpool.Conn) (count QueryCountResponse, err error) {
+	sql := `
+	SELECT COUNT(DISTINCT sender) FROM nft_event
+	WHERE action = 'new_class';
+	`
+	ctx, cancel := GetTimeoutContext()
+	defer cancel()
+
+	err = conn.QueryRow(ctx, sql).Scan(&count.Count)
+	if err != nil {
+		err = fmt.Errorf("get nft creator count failed: %w", err)
+		logger.L.Error(err)
+	}
+	return
+}
+
+func GetNftOwnerCount(conn *pgxpool.Conn) (count QueryCountResponse, err error) {
+	sql := `
+	SELECT COUNT(DISTINCT owner) FROM nft;
+	`
+	ctx, cancel := GetTimeoutContext()
+	defer cancel()
+
+	err = conn.QueryRow(ctx, sql).Scan(&count.Count)
+	if err != nil {
+		err = fmt.Errorf("get nft owner count failed: %w", err)
+		logger.L.Error(err)
+	}
+	return
+}
+
+func GetNftOwnerList(conn *pgxpool.Conn, p PageRequest) (res QueryNftOwnerListResponse, err error) {
+	sql := `
+	SELECT owner, COUNT(id) FROM nft
+	GROUP BY owner
+	ORDER BY COUNT(id) DESC
+	OFFSET $1
+	LIMIT $2
```

Comment:
Can we support filtering address/creator here somehow? Or is the caller expected to filter in application side?

Reply:
I think client side filtering is enough since when client side has the filter list, it can be done in the client side.